### ISSUE TITLE
Better user account merging logic

### ISSUE THREAD

**Issue by shayneczyzewski** [wasp-lang/wasp#954]
Our existing lookup process just checks if we already associated a user to the incoming social auth method/id and creates a new user if there is no association: https://github.com/wasp-lang/wasp/blob/main/waspc/data/Generator/templates/server/src/routes/auth/passport/generic/provider.js#L55

It would be better to allow an optional user-defined function to do this check. They could, for example, see that a GitHub user with email "shayne@wasp-lang.dev" is already associated with User 123, so when a Google user with the same email logs in for the first time we can associate them with the same user.

**Comment by shayneczyzewski** [comment]
It would also be nice if this method allowed users to link a social auth method to _their currently logged-in account_.

**Comment by Martinsos** [comment]
One thing to possibly keep an eye out for here is associating new social user with a email & password user that doesn't have a verified email -> that can result in a security issue where somebody created account with somebody's email and then gains access if that actual email owner logins with social account.

**Comment by Martinsos** [comment]
We have new issue for this, to actually re-implement it.

**Comment by infomiho** [comment]
@Martinsos I don't think we have a new issue for this? I used this one in our planning 😄

**Comment by Martinsos** [comment]
Ah whoops I thought you created a new one, something like "Implement account merging logic"! OK, leaving this one then!

**Comment by sodic** [comment]
User asking for this: https://discord.com/channels/686873244791210014/1271172858281852948/1271172858281852948

**Comment by PAD333** [comment]
> User asking for this: https://discord.com/channels/686873244791210014/1271172858281852948/1271172858281852948

Since you mention it: I also recently started using Wasp and this feature would be my nº1 priority.

**Comment by Joaquim-Ferrer** [comment]
This is a pretty important one for me too

**Comment by kouyaruten** [comment]
Also an important feature for me!

**Comment by sodic** [comment]
I'd also love seeing this. It would help Wasp set itself apart from other frameworks.

**Comment by sodic** [comment]
@JakeLo123 confirmed this is the biggest blocker for enabling social auth at their company (Searchcraft).

**Comment by dirkjbreeuwer** [comment]
+1 here ... Can't use social login without this.

**Comment by sachdva** [comment]
+1 here

**Comment by ricdays** [comment]
+1
Bumping this issue, as I just came across this issue - I find this missing feature a major disadvantage for this framework.

**Comment by sachdva** [comment]
+1 and in the meantime just some better error notifications would be really useful!

**Comment by rafalfigura** [comment]
+1 from me :)

**Comment by rachelslurs** [comment]
In the meantime, I'm wondering if there's a way to customize the messaging? "Save failed: user with the same identity already exists" might not be the most obvious to users saying "maybe you logged in using a different method", but I'm guessing saying something like that would be a security concern.

**Comment by Martinsos** [comment]
> In the meantime, I'm wondering if there's a way to customize the messaging? "Save failed: user with the same identity already exists" might not be the most obvious to users saying "maybe you logged in using a different method", but I'm guessing saying something like that would be a security concern.

We could probably do some improvements there without reducing security: I added an issue for that here! https://github.com/wasp-lang/wasp/issues/3002

**Comment by Case-E** [comment]
Came here to see if there was any progress here. I thought I added my vote on this one but I didn't. This is a blocker for us to enable social auth. We currently use email auth and once this is available in Wasp would like to add Google auth.

**Comment by sodic** [comment]
Thanks for the info @Case-E. We've been doing internal stuff during the last quarter - improving our processes, ci, and all that.

We'll soon get back to implementing user-facing features, and this one is likely the top priority.

**Comment by developerzeke** [comment]
Would it be worth switching to an auth library like `better-auth` to handle this kind of thing? It has [built in account linking](https://www.better-auth.com/docs/concepts/users-accounts#account-linking), provides a plugin ecosystem for more extensibility, and has 25+ oauth methods along with plugins for custom oauth and Anonymous, One-tap, and loads more sign-in options. I know a lot of work has gone in to building Wasp auth, but this could be better in the long run as it would be more flexible and have less code to maintain.

**Comment by Martinsos** [comment]
> Would it be worth switching to an auth library like `better-auth` to handle this kind of thing? It has [built in account linking](https://www.better-auth.com/docs/concepts/users-accounts#account-linking), provides a plugin ecosystem for more extensibility, and has 25+ oauth methods along with plugins for custom oauth and Anonymous, One-tap, and loads more sign-in options. I know a lot of work has gone in to building Wasp auth, but this could be better in the long run as it would be more flexible and have less code to maintain.

Yup we will be taking a look at it -> so one of big epics in the roadmap is to get Auth to netx level, and we will be als evaluation better-auth while doing it. Our hunch at the moment though is that it migth be too high-level for us, in the sense that we will have issues due t lack of control we can have over it. But we will know better once we evaluate it properly. In any case, good idea yes! less code to maintain is always great :D.

**Comment by infomiho** [comment]
Here's an example app that implements account merging using merge codes: https://github.com/wasp-lang/account-merging-example/

**Comment by sami-t** [comment]
+1 to switching to an auth library like Better Auth or possibly Stack Auth. Once you consider modern table stakes features like more providers, 2FA, passkeys, account merging, teams/orgs/projects, etc. the effort to integrate may be far less than the effort to maintain Wasp's own auth.

**Comment by simasima121** [comment]
Yep +1 for 2fa, account-merging etc.
